Source/ACE/Entity/Scroll.cs

```diff
@@ -0,0 +1,157 @@
+
+            switch (Power)
+            {
+                // research: http://asheron.wikia.com/wiki/Announcements_-_2002/06_-_Castling
+                case 50:  // Level 2
+                case 100: // Level 3
+                case 150: // Level 4
+                case 200: // Level 5
+                case 250: // Level 6
+                    if (session.Player.CheckMagicSkillLevel(School, Power))
+                        success = true;
+                    else
+                    {
+                        success = false;
+                        failReason = "You are not skilled enough in the inscribed spell's school of magic to understand the writing on this scroll.";
+                    }
+                    break;
+                default: // Level 1 or Level 7+ never fail
+                    success = true;
+                    break;
+            }
+
+            if (!session.Player.UnknownSpell(SpellId))
+            {
+                success = false;
+                failReason = "You already know the spell inscribed upon this scroll.";
+            }
+
+            ActionChain readScrollChain = new ActionChain();
+            readScrollChain.AddAction(session.Player, () => session.Player.HandleActionMotion(motionReading));
+            readScrollChain.AddDelaySeconds(2);
+
+            if (success)
+            {
+                readScrollChain.AddAction(session.Player, () => session.Player.HandleActionLearnSpell(SpellId));
+                readScrollChain.AddAction(session.Player, () => session.Player.HandleActionMotion(motionReady));
+                var removeObjMessage = new GameMessageRemoveObject(this);
+                var destroyMessage = new GameMessageSystemChat("The scroll is destroyed.", ChatMessageType.Magic);
+                readScrollChain.AddAction(session.Player, () => session.Network.EnqueueSend(destroyMessage, removeObjMessage));
+                readScrollChain.AddAction(session.Player, () => session.Player.RemoveFromInventory(Guid));
+            }
+            else
+            {
+                readScrollChain.AddDelaySeconds(2);
+                readScrollChain.AddAction(session.Player, () => session.Player.HandleActionMotion(motionReady));
+                var failMessage = new GameMessageSystemChat($"{failReason}", ChatMessageType.Magic);
+                readScrollChain.AddAction(session.Player, () => session.Network.EnqueueSend(failMessage));
+            }
+            var sendUseDoneEvent = new GameEventUseDone(session.Player.Session);
+            readScrollChain.AddAction(session.Player, () => session.Network.EnqueueSend(sendUseDoneEvent));
+            readScrollChain.EnqueueChain();
+        }
+
+        public override void SerializeIdentifyObjectResponse(BinaryWriter writer, bool success, IdentifyResponseFlags flags = IdentifyResponseFlags.None)
+        {          
+            var propertiesInt = PropertiesInt.Where(x => x.PropertyId == (uint)PropertyInt.Value
```

Comment:
a lot of this logic should be cached so we don't have to redo these lists every time.  An easy place to put this is the constructor, but you could inline it as well (but you'd have to make it threadsafe).